src/sockaddr.rs

```diff
@@ -168,34 +197,88 @@ impl SockAddr {
     }
 }
 
-// SocketAddrV4 and SocketAddrV6 are just wrappers around sockaddr_in and sockaddr_in6
-
-// check to make sure that the sizes at least match up
-fn _size_checks(v4: SocketAddrV4, v6: SocketAddrV6) {
-    unsafe {
-        mem::transmute::<SocketAddrV4, sockaddr_in>(v4);
-        mem::transmute::<SocketAddrV6, sockaddr_in6>(v6);
-    }
-}
-
 impl From<SocketAddrV4> for SockAddr {
     fn from(addr: SocketAddrV4) -> SockAddr {
-        unsafe {
-            SockAddr::from_raw_parts(
-                &addr as *const _ as *const _,
-                mem::size_of::<SocketAddrV4>() as socklen_t,
-            )
+        #[cfg(unix)]
+        let sin_addr = in_addr {
+            s_addr: u32::from_ne_bytes(addr.ip().octets()),
+        };
+        #[cfg(windows)]
+        let sin_addr = unsafe {
+            let mut s_un = mem::zeroed::<in_addr_S_un>();
+            *s_un.S_addr_mut() = u32::from_ne_bytes(addr.ip().octets());
+            in_addr { S_un: s_un }
+        };
+
+        let mut storage = MaybeUninit::<sockaddr_storage>::uninit();
```

Comment:
I'm thinking we should probably use `MaybeUninit::zeroed` to create these? Or possibly just `mem::zeroed()`. Currently after calling `assume_init` there are a lot of non-initialized bytes at the end of this `sockaddr_storage` struct.

Reply:
This file is currently a bit inconsistent. `SockAddr::from_raw_parts` use `MaybeUninit::<sockaddr_storage>::uninit()`, but `SockAddr::unix` uses `let mut addr = mem::zeroed::<sockaddr_un>();`

At least initializing all the structs to zero is probably a really good thing to do here. Whether or not directly with `mem::zeroed` or `MaybeUninit::zeroed` I don't have a strong opinion on.

Reply:
I had a comment that we likely needed to use zero: initialising only `sockaddr_in(6)` wasn't enough, as not the entire `sockaddr_storage` would be initialised.